### ISSUE TITLE
Issues with numbers less than 1 with Units and SigFigs

### ISSUE THREAD

**Issue by jmbragg** [maths/moodle-qtype_stack#232]
![sig fig bug stack](https://cloud.githubusercontent.com/assets/4606870/18036216/f2a9269c-6d32-11e6-801e-e51b3d458bd4.PNG)

See the attached picture.  You should be able to reproduce this with the metric conversions question attached here.
[questions-APChem-Factor Label-20160828-1514.zip](https://github.com/maths/moodle-qtype_stack/files/441309/questions-APChem-Factor.Label-20160828-1514.zip)

From what I've figured out so far this is only an issue for numbers much less that 1.  I've not identified where the cut point is at the moment.

**Comment by sangwinc** [comment]
Did you mean 1.81E-9Mg?  From your question variables you have

```
l1: [cg,mg,kg,ug]
l2: [1E-2,1E-3,1E3,1E-6]

l3: [ng,dg,Mg,pg,Gg]
l4: [1E-9,1E-1,1E6,1E-12,1E9]
```

This gives mg as 1E-3 and Mg as 1E6, so I think the conversion factor is 1E9?

In your questions you are not using "question tests".  These really help with reliable authoring!

**Comment by jmbragg** [comment]
No.  Its converting between cg and Mg.  So the conversion should be 1E8.

Yeah... I'm now convinced that I need to.  I got this one working late after the "g" function change and forgot to do it.

Thoughts otherwise?

**Comment by sangwinc** [comment]
Ok, sorry about the last silly comment!

I am, however confused. I hard-wired the randoms to reproduce your question.  Below is the result.

![issue232](https://cloud.githubusercontent.com/assets/781615/18062097/40f9d874-6e1d-11e6-9555-5a3a4caa0b95.png)

Please can you confirm which version of STACK you are using, and also which version of Maxima (including the precise LISP version)?

Thanks

**Comment by jmbragg** [comment]
Err... All the issues so far with old versions of Maxima seemed to be obscure and not relevant so I've been lazy and kept it at 5.23.2 on CentOS 6. The LISP is SBCL 1.0.38-3.el6.  Sadly, here is the first relevant one...

If that's what you think the issue is then I'll work on getting upgraded ASAP.

**Comment by jmbragg** [comment]
I've upgraded to 5.36.1 now.  The LISP is SBCL 1.0.38-3.el6 still.  The issue still occurs... See below.

It looks like its happening somewhere between 1E-6 and 1E-8.  The first example was a -8 that failed.  I just got a variant that was 1E-6 that worked as expected.

![sig fig bug stack 3](https://cloud.githubusercontent.com/assets/4606870/18111773/5454868e-6ef0-11e6-8eb3-de98777d5fc3.PNG)

**Comment by jmbragg** [comment]
Chris, I've been playing with this more since I upgraded the Maxima version.  I can verify that its not failing anymore for questions like the first one that I reported at 1E-8.  I've written a quick answer test and managed to deploy ones up to including 1E-9 pretty consistently.  I have some 1E-12's that will deploy but some do fail and I can't figure out what's different.  I have yet to get a 1E-15 or a 1E-11 to deploy but I don't know if its impossible.   

Any ideas?

**Comment by sangwinc** [comment]
Justin,

Can I ask you a favour?  Please could you start to create a list of
examples, a bit like this one.
https://github.com/maths/moodle-qtype_stack/blob/master/tests/fixtures/answertestfixtures.class.php

If you can create a list of example with what you do, and do not, think
should work we can start to investigate more fully what is going on.  These
lists are really useful in the long term, and will help track any future
regression in behaviour in future versions if we can fix this.

Chris

**Comment by jmbragg** [comment]
Will do.  I should have some time to play around with the variants to get examples this week.

Josh

**Comment by sangwinc** [comment]
Joshua,
If this issue is still live, please can I have some examples while we are working on issue #242?  Perhaps we can resolve these all together and release a new version?
Chris

**Comment by jmbragg** [comment]
Thanks for the reminder on this...  The frantic pace of teaching a course for the first time in 9 years...

I've uploaded a question that I wrote/modified to test this and an Excel sheet with most of the variants that I tried to deploy that failed.  Of interest in that spreadsheet are two values:
9.83e-10*cg - which failed
1.91e-10*cg - which passed
Everything less than 9.83e-10*cg failed.  Everything greater than 9.83e-10*cg passed.  I'm not saying that exact number is the break point but its somewhere in that neighborhood.  In fact, looking back at the bug, I evidently have some e-12's that passed on another version of a similar question.

A note, I'm not running the latest master here.  I think I have the latest in the plugins database.

[Problem Tests SigFigs.xlsx](https://github.com/maths/moodle-qtype_stack/files/531561/Problem.Tests.SigFigs.xlsx)
[questions-APChem-Bug Testing-20161015-2326.zip](https://github.com/maths/moodle-qtype_stack/files/531562/questions-APChem-Bug.Testing-20161015-2326.zip)
